README
PerlGuts Illustrated
--------------------

A collection of EPS illustrations that show the internal data
structures used by Perl. 
This stuff should also be available from <URL:http://search.cpan.org/dist/illguts/> as 
combined and extra html, extra pdf and combined chm.

Adapted to the two big refactorings at 
perl 5.10 (5.9.5) and 5.14 (5.13.2).

-------------------------------------------------------------------
Copyright 1998-1999 Gisle Aas.    All rights reserved.
Copyright 2009,2010,2012 Reini Urban.  All rights reserved.


Changes
0.49 2014-06-12 rurban
        * index-work.html (hv), hv-10.epsx:
          The NAME HEK for HV was added with 5.10 not 5.14. 5.14 just
          added the name_count hack. [RT #92650 by bulk88]
        * index-work.html (svrv): [RT #70723 by ikegami]
          No seperate RV type since 5.12

0.48 2014-06-12 rurban
        * gv-14.epsx: fix wrong ASSIGN_GEN hek

0.47 2013-12-11 rurban
        * svpv-8.epsx: add PVX ptr
        * svpv-{10,14}.epsx: add shared_hek pv
        * hv-18.epsx: added with 3 new xpvhv_aux fields
        * index-work.html, Makefile, htmlprep.pl: added index-18.html target.
        * index-work.html: explain shared strings via heks and COW,
        added a small cop_warnings par to OP

0.46 2013-10-01 rurban
        * magic.ps, magic-10.ps: fix /Helvtica typo which prevented
        svpvbm*.png being updated and cropped.
        * svpvbm-14.epsx: fix bbox and cur and len arrows
        * *.pdf: removed -12, used darwin to produce the pdfs for better images

0.45 2013-09-30 rurban
        * index: COW, COW_REFCNT
        added SV flags history
        * str-14.ps: added COW_REFCNT
        * svpv-14.epsx: updated bounding box

0.44 2013-08-23 rurban
        * index: update to 5.20 (SVt_INVLIST)
        * svtypes-14.epsx: added
        * svtypes.epsx: fixed wrong NV, PV arrows, removed BIND

0.43 2013-08-23 rurban
        * index: update to 5.18 (no structural changes done, refcounted PADLIST ignored so far)
        * pad.ps: fix @_ and NAMES confusion, thanks to LeoNerd

0.42 2012-04-09 rurban@cpan.org
        * index: spelling fixes
	* Makefile: added watch target (inotify)
	* META.yml: fix for make test_rel

0.41 2012-04-05 rurban@cpan.org
        * index: update IV,NV >= 5.10,
	  added arena
	* iv-10,14: updated

0.40 2012-04-04 rurban@cpan.org
        * index: finalize 5.16
	* hv-14: replace NAME chararray by hek
	* he: add flag to hek
	* gv-10,14: adjust for flag
	* strtab: use hek, remove myhek
	* eps2png: added sleep 0.01 for faster machines

0.39 2011-12-27 rurban@cpan.org
        * he, index: add HEK_FLAGS byte

0.38 2011-07-17 rurban@cpan.org
        * index: SvTYPE versions, lexical access
        * epsx2png: added
        * sviv-new.epsx: renamed to iv-typed.epsx

0.37 2011-07-03 rurban@cpan.org
        * index: REGEXP use just newSVpv
        * he.ps: yellow key in hek
        * cv-8: add PADLIST
	* hv-typed.ps, he-typed.ps, sviv-new.epsx, av-typed.epsx: 
	  added but yet unused

0.36 2011-01-08 rurban@cpan.org
        * index: REGEXP re_dup does regexp cloning
        add mjd gif from http://perl.plover.com/Rx/paper/

0.35 2010-11-05 rurban@cpan.org
        * hv-14: add name_count, remove FILL ptr

0.34 2010-09-11 rurban@cpan.org
        * svpvfm-14: do not crop LINES
	* added he-8.ps
	* mark index*.html files as read-only, generated.

0.33 2010-08-03 rurban@cpan.org
        enhanced the width of sv/xpv strings

        * he.ps: hek added from strtab
        * gv-10, gv-14: NAME_HEK use hek and not chararray
        * index: SvOOK_offset not IVX since 5.12
        * ook-12: added
        * ook-14: fixed offset, cur and len
        * io-8, hv-8: svpv-8, svpvmg-8, stack-8, strtab-8, sviv-8, svnv-8, svpvbm-8,
          svpviv-8, svpvlv-8, svpvnv-8, svpvrv-8, av-8, cv-8, gv-8: use sv-8.ps 
	  not sv.ps
        * scope-8, stack-8, he-8.ps: added
        * index-12.html: added for ook only
        * htmlprep.pl: add index-12.html for ook only

0.32 2010-07-22 rurban@cpan.org
        * io-14: LINES as IVX, no IFP

0.31 2010-07-21 rurban@cpan.org
        * av-14: fixed, removed # magic

0.30 2010-07-21 rurban@cpan.org
        * new 5.14 structs
	* added old and new versions: -8, -10 and -14 suffices
        * improved eps2png logic
        * moved index.html to index-work.html, create index*.html by htmlprep.pl

0.21 2010-03-30 rurban@cpan.org
        * index.html: C SvNULL typo.
        Added basic P5RX info.
        More links.

0.20 2010-02-10 rurban@cpan.org
       * index.html: 5.12 version bump.
	Mention self-ref'ed magic for REFCOUNTED

0.19 2010-01-06 rurban@cpan.org
	* stack: removed retstack, removed xpvav blocks, 
        use direct ARRAY ptr, smaller
	* context: new with PUSHBLOCK/PUSHSUB and PUSHSUBST
	* eval: new with PUSHBLOCK/PUSHEVAL and JMPENV
	* index.html: removed retstack. Added cxstack.
	Added sub and a eval chapter.
	Noted pp_ macros.
	* slides.pds: added 22) Context + sub, 23) eval
        * scope: added ENTER/LEAVE text to scopestack
	* hhk.pl: added
	* META.yml: added for a cpan/pause name
        * t/97_meta.t: added

0.18 2009-12-19 rurban@cpan.org
	* svpvgv: xnv_u => xbm_s, removed BmUSEFUL, BmPREVIOUS, BmRARE
	* index.html: xnv_u => xbm_s

0.17 2009-12-15 rurban@cpan.org
	* index.html: more links.
	  RV is not the exact same as a IV.
 	  PVBM changed to PVGV with VALID.

0.16 2009-12-01 rurban@cpan.org
	* opsamp2: added
	* eps2png: cleanup

0.15 2009-12-01 rurban@cpan.org
	* op.ps: type:9, opt:7
	* op1, op2: fixed cop,padop. removed condop
	* optypes: exchanged condop with padop

0.14 2009-11-30 rurban@cpan.org
	* op1.epsx: shortened pmop
	* pad.espx: added intermediate AVs, PL_curpad, PL_comppad
          and PL_comppad_name
	* svng,svpvbm,svpvmg: better bounding box
	* svpvrv.epsx, svpvrv.png: removed

0.13 2009-11-23 rurban@cpan.org
	* some typos
        * opsample.epsx: smaller bubble
	* Fixed opcode flags (parsed Concise vs static)
        * CV: added pad.ps with PADLIST ptr

0.12 2009-11-22 rurban@cpan.org
        * mk-slides: fix version
        * make_dist: force perms
        * slides/index.html: Removed unused S5 js from 
        * svpvmg.epsx: len,obj,ptr reshuffle, added magic_state
        * gv.epsx: renamed FLAGS to GvFLAGS
        * io.epsx: renamed FLAGS to IoFLAGS
        * index.html: Fixed GV
        * Added PAD to slides
	* Added opsample.epsx, added OP Concise explanation,
	  removed abc-simple.png

0.11 2009-11-22 rurban@cpan.org
        * Added Changes (written with info from backpan)

htmlprep.pl
#!/usr/bin/perl -w
use strict;
# split into version specific parts, and create expandable images for older versions
open I, "<", "index-work.html" or die;
my @HTML = qw(index.html index-8.html index-10.html index-12.html index-14.html index-18.html);
chmod 0644, @HTML;
open O, ">", "index.html" or die;
open O8, ">", "index-8.html" or die;
open O10, ">", "index-10.html" or die;
open O12, ">", "index-12.html" or die;
open O14, ">", "index-14.html" or die;
open O18, ">", "index-18.html" or die;
my %png = map { $_ => 1 } <*.png>;
my $from = qr/(?:<p>)?<center><img src="(.+)\.png"><\/center>/;

#my $to = '<div align="center"><span class="i58" id="svnull-8"><a name="svnull-8" onclick="javascript:Toggle(\'svnull-8\')"><span title="Click to hide">Until 5.8:</span><img class="i58" src="svnull-8.png" alt="svnull 5.8"></a></span><span class="i514" id="svhead"><a href="#svnull-8" onclick="javascript:Toggle(\'svnull-8\')"><span title="Click to show other">Since 5.10:<img src="svhead.png" alt="_SV_HEAD 5.10"></span></a></span></div>';

while (<I>) {
  my ($do58, $do510, $do512, $do514, $do518) = (0,0,0,0,0);
  if (/span.i58, span.i510 {display: none;}/) {
    print O;
    my $s = $_;
    $s =~ s/span.i58, span.i510 {display: none;}/span.i58, span.i510 {display: inline;}/;
    print O8  $s;
    print O10 $s;
    print O12 $s;
    print O14 $s;
    print O18 $s;
    next;
  }
  if (m|for perl 5.20 and older</small></h1>|) {
    print O;
    my $s = $_;
    $s =~ s/for perl 5.20 and older/for perl 5.8 and older/;
    print O8 $s;
    $s =~ s/for perl 5.8 and older/for perl 5.10/;
    print O10 $s;
    $s =~ s/for perl 5.10/for perl 5.12/;
    print O12 $s;
    $s =~ s/for perl 5.12/for perl 5.14 to 5.16/;
    print O14 $s;
    $s =~ s/for perl 5.14 to 5.16/since perl 5.18/;
    print O18 $s;
    next;
  }
  if (m|<title>PerlGuts Illustrated</title>|) {
    s|<title>PerlGuts Illustrated</title>|<title>PerlGuts Illustrated</title>\n<!-- Generated, do not edit! Edit index-work.html instead -->|;
  }
  if (m/ are now included in this single document, but also available/) {
    print O;
    my $s = $_;
    $s =~ s/ are now included in this single document, but also available/ are available/;
    print O8 $s;
    print O10 $s;
    print O12 $s;
    print O14 $s;
    print O18 $s;
    next;
  }
  if (/$from/) {
    my $png = $1;
    $do58++  if $png{$png."-8.png"};
    $do510++ if $png{$png."-10.png"};
    $do512++ if $png{$png."-12.png"};
    $do514++ if $png{$png."-14.png"};
    $do518++ if $png{$png."-18.png"};
    if ($png{$png.".png"} and !($do58 or $do510 or $do514 or $do518)) {
      s/$from/<center><img src="$png\.png" alt="$png"><\/center>/; # keep it, just add alt=$png
      print O8;
      print O10;
      print O12;
      print O14;
      print O18;
    } else {
      $_ = '<div align="center">';
      if ($do58 and $do510 and $do514) {
        print O8 $_."<img class=\"i58\" src=\"$png-8.png\" alt=\"$png 5.8\">"."</div>\n";
        print O10 $_."<img class=\"i510\" src=\"$png-10.png\" alt=\"$png 5.10\">"."</div>\n";
        print O12 $_."<img class=\"i514\" src=\"$png-".($do512?"12":"10").".png\" alt=\"$png 5.12\">"."</div>\n" if $do512 or $do510;
        print O14 $_."<img class=\"i514\" src=\"$png-14.png\" alt=\"$png 5.14\">"."</div>\n";
        print O18 $_."<img class=\"i518\" src=\"$png-18.png\" alt=\"$png 5.18\">"."</div>\n" if $do518;
        print O18 $_."<img class=\"i518\" src=\"$png-14.png\" alt=\"$png 5.14\">"."</div>\n" unless $do518;

        $_ .= "<span class=\"i58\" id=\"$png-8\"><a name=\"$png-8\" onclick=\"javascript:Toggle(\'$png-8\')\" title=\"Click to hide\">Until 5.8:<img class=\"i58\" src=\"$png-8.png\" alt=\"$png 5.8\"><br>".
          "Since 5.10:<img class=\"i510\" src=\"$png-10.png\" alt=\"$png 5.10\"></a></span><br>".
          "<span class=\"i514\" id=\"$png-14\"><a name=\"$png-14\" href=\"#$png-8\" onclick=\"javascript:Toggle(\'$png-8\')\" title=\"Click to show other\">Since 5.14:<img class=\"i514\" src=\"$png-14.png\" alt=\"$png 5.14\"></a></span>";
        $_ .= "<span class=\"i518\" id=\"$png-18\"><a name=\"$png-18\" href=\"#$png-8\" onclick=\"javascript:Toggle(\'$png-8\')\" title=\"Click to show other\">Since 5.18:<img class=\"i518\" src=\"$png-18.png\" alt=\"$png 5.18\"></a></span>" if $do518;

      } else {
        print O8 $_."<img class=\"i58\" src=\"$png-8.png\" alt=\"$png 5.8\">"."</div>\n"     if $do58;
        print O10 $_."<img class=\"i510\" src=\"$png-10.png\" alt=\"$png 5.10\">"."</div>\n" if $do510;
        print O12 $_."<img class=\"i514\" src=\"$png-".($do512?"12":"10").".png\" alt=\"$png 5.12\">"."</div>\n" if $do512 or $do510;
        print O14 $_."<img class=\"i514\" src=\"$png-14.png\" alt=\"$png 5.14\">"."</div>\n" if $do514;
        print O18 $_."<img class=\"i518\" src=\"$png-18.png\" alt=\"$png 5.18\">"."</div>\n" if $do518;
        print O18 $_."<img class=\"i518\" src=\"$png-14.png\" alt=\"$png 5.14\">"."</div>\n" if !$do518 and $do514;

        print O8 $_."<img class=\"i58\" src=\"$png.png\" alt=\"$png\">"."</div>\n"   if !$do58 and $png{$png.".png"};
        print O10 $_."<img class=\"i510\" src=\"$png.png\" alt=\"$png\">"."</div>\n" if !$do510 and $png{$png.".png"};
        print O12 $_."<img class=\"i510\" src=\"$png.png\" alt=\"$png\">"."</div>\n" if !$do512 and $png{$png.".png"};
        print O14 $_."<img class=\"i514\" src=\"$png.png\" alt=\"$png\">"."</div>\n" if !$do514 and $png{$png.".png"};
        print O18 $_."<img class=\"i518\" src=\"$png.png\" alt=\"$png\">"."</div>\n" if !$do518 and $png{$png.".png"};
        # cases: 58+. (new)/58+none (svpvrv)/58+510+514 (handled above)

        # old hidden
        $_ .= "<span class=\"i58\" id=\"$png-8\"><a name=\"$png-8\" onclick=\"javascript:Toggle(\'$png-8\')\" title=\"Click to hide\">Until 5.8:<img class=\"i58\" src=\"$png-8.png\" alt=\"$png 5.8\"></a></span><br>" if $do58;
        $_ .= "<span class=\"i510\" id=\"$png-10\"><a name=\"$png-10\" onclick=\"javascript:Toggle(\'$png-10\')\" title=\"Click to hide\">Since 5.10:<img class=\"i510\" src=\"$png-10.png\" alt=\"$png 5.10\"></a></span><br>" if ($do510 and $do514);
        $_ .= "<span class=\"i514\" id=\"$png-12\"><a name=\"$png-12\" onclick=\"javascript:Toggle(\'$png-12\')\" title=\"Click to hide\">Since 5.12:<img class=\"i514\" src=\"$png-12.png\" alt=\"$png 5.12\"></a></span><br>" if $do512; # ook only
        # new visible, there's always a -8 version
        $_ .= "<span class=\"i514\" id=\"$png-14\"><a name=\"$png-14\" onclick=\"javascript:Toggle(\'$png-8\')\" title=\"Click to show other\">Since 5.14:<img class=\"i514\" src=\"$png-14.png\" alt=\"$png 5.14\"></a></span>" if $do514;
        $_ .= "<span class=\"i518\" id=\"$png-18\"><a name=\"$png-18\" onclick=\"javascript:Toggle(\'$png-8\')\" title=\"Click to show other\">Since 5.18:<img class=\"i518\" src=\"$png-18.png\" alt=\"$png 5.18\"></a></span>" if $do518;
        $_ .= "<span id=\"$png\"><a name=\"$png\" href=\"#$png-8\" onclick=\"javascript:Toggle(\'$png-8\')\" title=\"Click to show other\">Since 5.10:<img class=\"i514\" src=\"$png.png\" alt=\"$png\"></a></span>" if $do58 and !$do510 and !$do514 and !$do518 and $png{$png.".png"};
      }
      $_ .= "</div>\n";
    }
  } else {
    print O8;
    print O10;
    print O12;
    print O14;
    print O18;
  }
  print O;
}
close I;
chmod 0444, @HTML;
